src/Evolution/EventsAndDenseTriggers/EventsAndDenseTriggers.hpp

```diff
@@ -0,0 +1,247 @@
+  };
+
+  // The data structure used here is a heap containing the triggers
+  // not being tested at the moment, with the earliest time as the
+  // root at index 0, followed by the triggers known to trigger at the
+  // current time, followed by the triggers currently being processed.
+  // When we are done with the current time, we choose the next time
+  // to process by examining the heap root and then pop all triggers
+  // requesting that time into the processing area.  Triggers are
+  // moved back to the heap either when we know they don't trigger or
+  // when we have run their events.
+  //
+  // Except at initialization (when everything needs to be processed
+  // at the start time), the trigger processing area is expected to be
+  // small.
+  Storage events_and_triggers_;
+  // The size of the heap, or equivalently the index of the start of
+  // the processing area.  The initial -1 is a sentinel for an
+  // uninitialized state.
+  typename Storage::difference_type heap_size_ = -1;
+  // Index of the next trigger to process.
+  typename Storage::difference_type processing_position_{};
+  double next_check_ = std::numeric_limits<double>::signaling_NaN();
+  TriggerTimeAfter next_check_after_{};
+};
+
+template <typename DbTags>
+double EventsAndDenseTriggers::next_trigger(
+    const db::DataBox<DbTags>& box) noexcept {
+  if (UNLIKELY(heap_size_ == -1)) {
+    initialize(box);
+  }
+
+  if (events_and_triggers_.empty()) {
+    return next_check_after_.infinite_future();
+  }
+
+  return next_check_;
+}
+
+template <typename DbTags, typename Metavariables, typename ArrayIndex,
+          typename ComponentPointer>
+EventsAndDenseTriggers::TriggeringState EventsAndDenseTriggers::is_ready(
+    const db::DataBox<DbTags>& box, Parallel::GlobalCache<Metavariables>& cache,
+    const ArrayIndex& array_index, const ComponentPointer component) noexcept {
+  ASSERT(heap_size_ != -1, "Not initialized");
+  ASSERT(not events_and_triggers_.empty(),
+         "Should not be calling is_ready with no triggers");
+  ASSERT(heap_end() != events_and_triggers_.end(), "No active triggers");
+
+  const evolution_greater<double> after{
+      db::get<::Tags::TimeStepId>(box).time_runs_forward()};
+
+  while (current_trigger() != events_and_triggers_.end()) {
+    if (not current_trigger()->trigger->is_ready(box)) {
+      return TriggeringState::NotReady;
+    }
+
+    const auto is_triggered = current_trigger()->trigger->is_triggered(box);
+    if (not after(is_triggered.next_check, current_trigger()->next_check)) {
+      ERROR("Trigger at time " << current_trigger()->next_check
+            << " rescheduled itself for earlier time "
+            << is_triggered.next_check);
+    }
+    current_trigger()->next_check = is_triggered.next_check;
+    if (not is_triggered.is_triggered) {
+      finish_processing_trigger_at_current_time(current_trigger());
+    }
+    ++processing_position_;
+  }
+
+  for (auto trigger = heap_end();
```

Comment:
sorry, I don't understand why this is `heap_end()`. Do you think it would be helpful to add a C++ comment here to clarify? I expect that if I understood why `heap_end()` is used here I'd understand below as well